README.rst
Jupyter MicroPython Upydevice Kernel
=====================================

Jupyter kernel to interact with a MicroPython board over its REPL
interface.

Typically used with MicroPython boards over the USB / Serial interface,
or through the WebREPL.

MicroPython
-----------
Tested on MicroPython v1.11 - v1.16

Boards
-------
Tested on:
  - PYBOARD V1.1/Lite
  - ESP32
  - ESP8266
  - CIRCUITPLAYGROUND EXPRESS


IDE/PLATFORMS
-------------
- Jupyter notebooks
- Jupyter lab
- Atom (`Hydrogen <https://github.com/nteract/hydrogen>`_)
- Visual Studio Code (`Jupyter notebooks <https://code.visualstudio.com/docs/python/jupyter-support>`_)

Background
----------

This Jupyter MicroPython Kernel is heavily based on the amazing work
done on https://github.com/goatchurchprime/jupyter_micropython_kernel
and https://github.com/andrewleech/jupyter_micropython_remote

Their device connection library has been replaced by
upydevice latest classes ``SerialDevice``, ``WebSocketDevice`` and ``AsyncBleDevice``
that allows Serial, WiFi (websocket-WebREPL) and Bluetooth Low Energy (BleREPL) connections.
The kernel has also been reworked to support autocompletions on tab which works
for MicroPython, iPython and %cell magic commands.
Some %cell magic commands were dropped and some new were added e.g: ``%is_reachable``
``%meminfo`` ``%whoami`` ``%gccollect`` ``%sync`` ``%logdata`` ``%devplot`` ``%connect``
``%rssi``

Installation
------------

Ensure you have a new enough MicroPython installed on your board (see above).

You also need python 3.6 or above, ensuring it’s available from your current
command line. Optionally (recommended) use your favourite pipenv / virtualenv to set
up a clean environment to run jupyter from.

Then install this module:

::

   pip install jupyter-micropython-upydevice

Install the kernel into jupyter itself using the shell command:

::

   python -m mpy_kernel_upydevice.install

This registers the kernel with Jupyter so it can be selected for use in
notebooks

Running
-------

Now run Jupyter notebooks or Jupyter lab:

For Jupyter lab do:

::

   jupyter lab

In the launcher click on the MicroPython upydevice button to create a new notebook

Serial device
-------------

To connect through serial/USB: ``%serialconnect [SERIAL PORT] [BAUDRATE]``
or if a device already configured (see `upydev <https://github.com/Carglglz/upydev>`_)
in the global group 'UPY_G',  ``%serialconnect @[DEVICE]`` which has autocompletion
on tab.

This command has autocompletion on tab, so hit tab and select a port from the list

Baudrate default is 115200

*Use -kbi option to interrupt any running loop*

eg:
::

   %serialconnect /dev/tty.usbmodem3370377430372 115200



or

::

    %serialconnect @pybV1.1

::

  ** Serial connected **

  SerialDevice @ /dev/tty.usbmodem3370377430372, Type: pyboard, Class: SerialDevice
  Firmware: MicroPython v1.17-217-gde7e3cd79-dirty on 2021-11-30; PYBv1.1 with STM32F405RG
  Pyboard Virtual Comm Port in FS Mode, Manufacturer: MicroPython
  (MAC: 3c:00:3d:00:02:47:37:30:38:37:33:33)

  MicroPython v1.17-217-gde7e3cd79-dirty on 2021-11-30; PYBv1.1 with STM32F405RG
  Type "help()" for more information.


Wireless Device: WiFi
---------------------

To connect through WebREPL: ``%websocketconnect [IP] --password "[PASSWORD]"``
or if a device already configured (see `upydev <https://github.com/Carglglz/upydev>`_)
in the global group 'UPY_G',  ``%websocketconnect @[DEVICE]`` which has autocompletion
on tab.

*Use -kbi option to interrupt any running loop*

eg:

::

    %websocketconnect 192.168.1.42 --password "mypass"



or

::

    %websocketconnect @esp_room1

::

  ** WebREPL connected **

  WebSocketDevice @ ws://192.168.1.66:8266, Type: esp32, Class: WebSocketDevice
  Firmware: MicroPython v1.17 on 2021-09-02; ESP32 module with ESP32
  (MAC: 7c:9e:bd:3d:9d:f4, Host Name: gkesp32, RSSI: -58 dBm)

  MicroPython v1.17 on 2021-09-02; ESP32 module with ESP32
  Type "help()" for more information.


Wireless Device: BLE
--------------------

To connect through BLEREPL: ``%bleconnect [ADDRESS] "``
or if a device already configured (see `upydev <https://github.com/Carglglz/upydev>`_)
in the global group 'UPY_G',  ``%bleconnect @[DEVICE]`` which has autocompletion
on tab.


eg:

::

    %bleconnect 9998175F-9A91-4CA2-B5EA-482AFC3453B9



or

::

    %bleconnect @bledev

::

  [Service] 6e400001-b5a3-f393-e0a9-e50e24dcca9e: Nordic UART Service
  [Characteristic] 6e400003-b5a3-f393-e0a9-e50e24dcca9e: (notify) | Name: Nordic UART TX
    [Descriptor] [00002902-0000-1000-8000-00805f9b34fb]: Client Characteristic Configuration (Handle: 22)
  [Characteristic] 6e400002-b5a3-f393-e0a9-e50e24dcca9e: (write) | Name: Nordic UART RX

  ** BleREPL connected **

  BleDevice @ 9998175F-9A91-4CA2-B5EA-482AFC3453B9, Type: esp32 , Class: BleDevice
  Firmware: MicroPython v1.16 on 2021-06-24; ESP32 module with ESP32
  (MAC: 30:ae:a4:23:35:64, Local Name: esp32-34, RSSI: -50 dBm)

  MicroPython v1.16 on 2021-06-24; ESP32 module with ESP32
  Type "help()" for more information.


The ``%connect`` magic command allows any type of device:
::

    %connect @esp_room1

::

    %connect @pybV1.1

::

    %connect @bledev



You should now be able to execute MicroPython commands by running the
cells.

There is a *upydevice_kernel_demo.ipynb* file in the directory you could look at
with some of the features shown.

If a cell is taking too long, or if you want to stop an ongoing loop
To interrupt / ^C use “Interrupt the kernel” button and this will raise a ``KeyboardInterrupt``
in the device.

Alternatively hit Escape and then ‘i’ twice.

To do a soft reset (it will reconnect automatically) type:

::

   %rebootdevice

  Note: Restarting the kernel does not actually reboot the device.
  Also, pressing the reset button will mess things up (at least for WebREPL, BleREPL
  and for serial this is probably true as well).

%Cell magic Commands
--------------------
Cell magic commands have autocompletion on tab, so hit tab and see what Commands
are available, or to see more information about each command do:
::

   %lsmagic


::

  %disconnect
    disconnects device

  %lsmagic
      list magic commands

  %rebootdevice
      reboots device

  %is_reachable
      Test if device is reachable (must be connected first)

  %connect
      connects to a device based on addres or configuration

  %serialconnect [-kbi] [portname] [baudrate]
      connects to a device over USB, default baudrate is 115200

  %websocketconnect [--password PASSWORD] [-kbi] [-ssl] [websocketurl]
      connects to the WebREPL over wifi (WebREPL daemon must be running)
      websocketurl defaults to 192.168.4.1 (uri -> ws://192.168.4.1:8266)

  %bleconnect [bleaddress]
      connects to the BleREPL over Bluetooth Low Energy(BleREPL must be running)

  %meminfo
      Shows RAM size/used/free/use% info

  %whoami
      Shows Device name, port, id, and system info

  %rssi
      Shows Device RSSI if wireless

  %gccollect
      To use the garbage collector and free some RAM if possible

  %local
      To run the cell contents in local iPython

  %sync
      To sync a variable/output data structure of the device into iPython
      if no var name provided it stores the output into _

  %logdata [-fs FS] [-tm TM] [-u U [U ...]] [-s] v [v ...]
      To log output data of the device into iPython,
      data is stored in 'devlog'

     positional arguments:
        v             Name of variables
     optional arguments:
        -fs FS        Sampling frequency in Hz
        -tm TM        Sampling timeout in ms
        -u U [U ...]  Unit of variables
        -s            Silent mode

  %devplot
      To plot devlog data

The communications interface to the MicroPython device is based on `upydevice
<https://github.com/Carglglz/upydevice>`


This is also the core library of `upydev
<https://github.com/Carglglz/upydev>`_ .
The SERIAL SHELL-REPL can be used simultaneously
with the upydevice Kernel since the serial connection is non-blocking.

%local
------
Individual cells can also be run on the local iPython instead of the MicroPython
kernel by starting a cell with ``%local``

This can be useful to work directly with local files, use ipywidgets, etc.
Commands here will be run by the standard ipython kernel.

In `%local` cells, a special global function ``remote()`` is also available which
will pass a single string argument to the MicroPython board to be run, returning
any stdout from the command. Eg:

micropython cell

::

   from machine import Pin
   import neopixel
   pixels = neopixel.NeoPixel(Pin(4, Pin.OUT), 1)

   def set_colour(r, g, b):
       pixels[0] = (r, g, b)
       pixels.write()

   set_colour(0xff, 0xff, 0xff)

local cell

::

   %local
   import colorsys
   from ipywidgets import interact, Layout, FloatSlider

   def set_hue(hue):
       r, g, b = (int(p*255) for p in colorsys.hsv_to_rgb(hue, 1.0, 1.0))
       remote(f"set_colour({r}, {g}, {b})")

   slider = FloatSlider(min=0,max=1.0,step=0.01, layout=Layout(width='80%', height='80px'))
   interact(set_hue, hue=slider)



\



%sync
-----

Any variable/output of the device can be stored in local iPython easily.
If a var name is not provided the output will be stored locally in _ , e.g. :

::

    %sync
    [1,2,3,5]

::

  [1, 2, 3, 5]


::

    %local
    _

::

  [1, 2, 3, 5]


If device output is assigned to a variable it will be stored locally with the
same name e.g. :

::

    %sync
    my_data = {'key1':[1,2,3], 'key2':[4,5,6]}


::

    %local
    my_data


::

    {'key2': [4, 5, 6], 'key1': [1, 2, 3]}





This works for any type of output (bytes/bytearrays/arrays/ints/floats/strings/lists/dicts)

%logdata
---------
This allows to log any data from device stdout as long as the data is in tuple or list format.
The data will be stored in local iPython in 'devlog'. *(Serial and WebSocketDevices only)*

positional arguments:
   v             Name of variables
optional arguments:
   - -fs FS        Sampling frequency in Hz
   - -tm TM        Sampling timeout in ms
   - -u U [U ...]  Unit of variables
   -  \-s            Silent mode

e.g. :
Logging accelerometer data from an IMU sensor.

*micropython cell*

::

    import time
    from machine import I2C, Pin
    from lsm9ds1 import LSM9DS1
    i2c = I2C(scl=Pin(22), sda=Pin(23))
    imu = LSM9DS1(i2c)

    def stream_accel(n, tm):
      for i in range(n):
          print(imu.read_accel())
          time.sleep_ms(tm)


*%logdata cell*

::

    %logdata 'x' 'y' 'z' -tm 10 -u 'g(9.8m/s^2)'
    stream_accel(400, 10)

::

    vars:['x', 'y', 'z'], fs:None Hz, tm:10 ms, u: ['g(9.8m/s^2)'], silent: False
    ------------------------------
    (-0.6851807, 0.6947632, 0.3374634)
    (-0.6889038, 0.6830444, 0.3411255)
    (-0.7027588, 0.6877441, 0.3455811)
    (-0.7280884, 0.7080688, 0.3401489)
    ....
    (-0.734375, 0.7600098, -0.0004272461)
    (-0.7210693, 0.7717896, -0.05194092)
    (-0.7344971, 0.7575684, 0.006652832)


Now data is stored in devlog

::

    %local
    devlog

::

    {'x': [-0.6851807, ..., -0.7344971], 'y': [0.6947632, ..., 0.7575684],
     'z': [-0.7280884, ..., 0.006652832], 'vars': ['x', 'y', 'z']
     'fs': 100, 'ts': [0.0, ... , 4.0], 'u': ['g(9.8m/s^2)']}


%devplot
--------
This allows to plot *devlog* data, just do:


::

    %devplot



.. image:: acc-plot.png
    :width: 2140pt


Now to save the plot do:

::

    %local
    fig.savefig('acc-plot.png')


::

    %local
    %ls

::

    LICENSE*                    mpy_kernel_upydevice/
    README.rst*                 setup.py
    acc-plot.png                upydevie_kernel_demo.ipynb


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.8] [Github repo]
## [0.0.7] - 2022-03-04
### Fix
- autocompletion crash if no device connected
- debug info if device is not reachable
## [0.0.6] - 2021-12-18
### Added
- `%connect` magic command to connect to a device based on device type
- `%rssi` magic command to get RSSI of the device (WebSocketDevice, and BleDevice)
### Fix
- Autocompletion of configured devices
- Handle on connection error, not configured or not available.
- Disconnection status
## [0.0.5] - 2021-08-15
### Added
- Bledevice
## [0.0.4] - 2020-05-25
### Fix
- Serial is reachable fix
## [0.0.3] - 2020-02-24
### Added
- Monokai style to jupyter console
## [0.0.2] - 2020-02-17
### Fix
- esp8266 WS_DEVICE fix
## [0.0.1] - 2020-02-16
### Added
- First release


mpy_kernel_upydevice/install.py
import json
import os
import sys
import argparse

from jupyter_client.kernelspec import KernelSpecManager
from IPython.utils.tempdir import TemporaryDirectory

# copied out from https://github.com/takluyver/bash_kernel/blob/master/bash_kernel/install.py

# sys.executable should be "python3"
kernel_json = {"argv": [sys.executable, "-m", "mpy_kernel_upydevice", "-f", "{connection_file}"],
               "display_name": "MicroPython upydevice kernel",
               "language": "python"
               }


def install_my_kernel_spec(user=True, prefix=None):
    if "python2" in sys.executable:
        print("I think this needs python3")
    with TemporaryDirectory() as td:
        os.chmod(td, 0o755) # Starts off as 700, not user readable
        with open(os.path.join(td, 'kernel.json'), 'w') as f:
            json.dump(kernel_json, f, sort_keys=True)
        # TODO: Copy resources once they're specified

        print('Installing IPython kernel spec of micropython')
        k = KernelSpecManager()
        k.install_kernel_spec(td, 'Micropython-upydevice', user=user, prefix=prefix)

        h = k.get_kernel_spec("micropython-upydevice")
        print("...into", h.resource_dir)


def _is_root():
    try:
        return os.geteuid() == 0
    except AttributeError:
        return False # assume not an admin on non-Unix platforms

def main(argv=None):
    parser = argparse.ArgumentParser(
        description='Install KernelSpec for MicroPython Kernel'
    )
    prefix_locations = parser.add_mutually_exclusive_group()

    prefix_locations.add_argument(
        '--user',
        help='Install KernelSpec in user homedirectory',
        action='store_true'
    )
    prefix_locations.add_argument(
        '--sys-prefix',
        help='Install KernelSpec in sys.prefix. Useful in conda / virtualenv',
        action='store_true',
        dest='sys_prefix'
    )
    prefix_locations.add_argument(
        '--prefix',
        help='Install KernelSpec in this prefix',
        default=None
    )

    args = parser.parse_args(argv)

    user = False
    prefix = None
    if args.sys_prefix:
        prefix = sys.prefix
    elif args.prefix:
        prefix = args.prefix
    elif args.user or not _is_root():
        user = True

    install_my_kernel_spec(user=user, prefix=prefix)

if __name__ == '__main__':
    main()


mpy_kernel_upydevice/set_monokai.py
#!/usr/bin/env python3

import os
from mpy_kernel_upydevice import __path__
import shutil

jupyter_config_dir = '.jupyter'
jupyter_config_file = 'jupyter_console_config.py'


def main():

    if jupyter_config_dir not in os.listdir(os.environ['HOME']):
        os.mkdir("{}/{}".format(os.environ['HOME'], jupyter_config_dir))
    else:
        print('Found existing .jupyter config directory')

    print('Setting monokai as default jupyter console style')

    try:
        shutil.copyfile('{}/{}'.format(__path__[0], jupyter_config_file),
                        '{}/{}/{}'.format(os.environ['HOME'], jupyter_config_dir,
                                          jupyter_config_file))
        print('Monokai style setup successful!')
    except Exception as e:
        print(e)


if __name__ == '__main__':
    main()


mpy_kernel_upydevice/jupyter_console_config.py
# Configuration file for jupyter-console.

# ------------------------------------------------------------------------------
# ConnectionFileMixin(LoggingConfigurable) configuration
# ------------------------------------------------------------------------------

# Mixin for configurable classes that work with connection files

# JSON file in which to store connection info [default: kernel-<pid>.json]
#
#  This file will contain the IP, ports, and authentication key needed to connect
#  clients to this kernel. By default, this file will be created in the security
#  dir of the current profile, but can be specified by absolute path.
#c.ConnectionFileMixin.connection_file = ''

# set the control (ROUTER) port [default: random]
#c.ConnectionFileMixin.control_port = 0

# set the heartbeat port [default: random]
#c.ConnectionFileMixin.hb_port = 0

# set the iopub (PUB) port [default: random]
#c.ConnectionFileMixin.iopub_port = 0

# Set the kernel's IP address [default localhost]. If the IP address is
#  something other than localhost, then Consoles on other machines will be able
#  to connect to the Kernel, so be careful!
#c.ConnectionFileMixin.ip = ''

# set the shell (ROUTER) port [default: random]
#c.ConnectionFileMixin.shell_port = 0

# set the stdin (ROUTER) port [default: random]
#c.ConnectionFileMixin.stdin_port = 0

##
#c.ConnectionFileMixin.transport = 'tcp'

# ------------------------------------------------------------------------------
# JupyterConsoleApp(ConnectionFileMixin) configuration
# ------------------------------------------------------------------------------

# Set to display confirmation dialog on exit. You can always use 'exit' or
#  'quit', to force a direct exit without any confirmation.
#c.JupyterConsoleApp.confirm_exit = True

# Connect to an already running kernel
#c.JupyterConsoleApp.existing = ''

# The name of the default kernel to start.
#c.JupyterConsoleApp.kernel_name = 'python'

# Path to the ssh key to use for logging in to the ssh server.
#c.JupyterConsoleApp.sshkey = ''

# The SSH server to use to connect to the kernel.
#c.JupyterConsoleApp.sshserver = ''

# ------------------------------------------------------------------------------
# Application(SingletonConfigurable) configuration
# ------------------------------------------------------------------------------

# This is an application.

# The date format used by logging formatters for %(asctime)s
#c.Application.log_datefmt = '%Y-%m-%d %H:%M:%S'

# The Logging format template
#c.Application.log_format = '[%(name)s]%(highlevel)s %(message)s'

# Set the log level by value or name.
#c.Application.log_level = 30

# ------------------------------------------------------------------------------
# JupyterApp(Application) configuration
# ------------------------------------------------------------------------------

# Base class for Jupyter applications

# Answer yes to any prompts.
#c.JupyterApp.answer_yes = False

# Full path of a config file.
#c.JupyterApp.config_file = ''

# Specify a config file to load.
#c.JupyterApp.config_file_name = ''

# Generate default config file.
#c.JupyterApp.generate_config = False

# ------------------------------------------------------------------------------
# ZMQTerminalIPythonApp(JupyterApp,JupyterConsoleApp) configuration
# ------------------------------------------------------------------------------

# ------------------------------------------------------------------------------
# ZMQTerminalInteractiveShell(SingletonConfigurable) configuration
# ------------------------------------------------------------------------------

# Text to display before the first prompt. Will be formatted with variables
#  {version} and {kernel_banner}.
#c.ZMQTerminalInteractiveShell.banner = 'Jupyter console {version}\n\n{kernel_banner}'

# Callable object called via 'callable' image handler with one argument, `data`,
#  which is `msg["content"]["data"]` where `msg` is the message from iopub
#  channel.  For example, you can find base64 encoded PNG data as
#  `data['image/png']`. If your function can't handle the data supplied, it
#  should return `False` to indicate this.
#c.ZMQTerminalInteractiveShell.callable_image_handler = None

# Shortcut style to use at the prompt. 'vi' or 'emacs'.
#c.ZMQTerminalInteractiveShell.editing_mode = 'emacs'

# Highlight matching brackets.
#c.ZMQTerminalInteractiveShell.highlight_matching_brackets = True

# The name of a Pygments style to use for syntax highlighting
c.ZMQTerminalInteractiveShell.highlighting_style = 'monokai'

# Override highlighting format for specific tokens
#c.ZMQTerminalInteractiveShell.highlighting_style_overrides = {}

# How many history items to load into memory
#c.ZMQTerminalInteractiveShell.history_load_length = 1000

# Handler for image type output.  This is useful, for example, when connecting
#  to the kernel in which pylab inline backend is activated.  There are four
#  handlers defined.  'PIL': Use Python Imaging Library to popup image; 'stream':
#  Use an external program to show the image.  Image will be fed into the STDIN
#  of the program.  You will need to configure `stream_image_handler`;
#  'tempfile': Use an external program to show the image.  Image will be saved in
#  a temporally file and the program is called with the temporally file.  You
#  will need to configure `tempfile_image_handler`; 'callable': You can set any
#  Python callable which is called with the image data.  You will need to
#  configure `callable_image_handler`.
#c.ZMQTerminalInteractiveShell.image_handler = 'PIL'

# Whether to include output from clients other than this one sharing the same
#  kernel.
#
#  Outputs are not displayed until enter is pressed.
#c.ZMQTerminalInteractiveShell.include_other_output = False

# Timeout (in seconds) for giving up on a kernel's is_complete response.
#
#  If the kernel does not respond at any point within this time, the kernel will
#  no longer be asked if code is complete, and the console will default to the
#  built-in is_complete test.
#c.ZMQTerminalInteractiveShell.kernel_is_complete_timeout = 1

# Timeout for giving up on a kernel (in seconds).
#
#  On first connect and restart, the console tests whether the kernel is running
#  and responsive by sending kernel_info_requests. This sets the timeout in
#  seconds for how long the kernel can take before being presumed dead.
#c.ZMQTerminalInteractiveShell.kernel_timeout = 60

# Preferred object representation MIME type in order.  First matched MIME type
#  will be used.
#c.ZMQTerminalInteractiveShell.mime_preference = ['image/png', 'image/jpeg', 'image/svg+xml']

# Prefix to add to outputs coming from clients other than this one.
#
#  Only relevant if include_other_output is True.
#c.ZMQTerminalInteractiveShell.other_output_prefix = '[remote] '

# Use simple fallback prompt. Features may be limited.
#c.ZMQTerminalInteractiveShell.simple_prompt = False

# Command to invoke an image viewer program when you are using 'stream' image
#  handler.  This option is a list of string where the first element is the
#  command itself and reminders are the options for the command.  Raw image data
#  is given as STDIN to the program.
#c.ZMQTerminalInteractiveShell.stream_image_handler = []

# Command to invoke an image viewer program when you are using 'tempfile' image
#  handler.  This option is a list of string where the first element is the
#  command itself and reminders are the options for the command.  You can use
#  {file} and {format} in the string to represent the location of the generated
#  image file and image format.
#c.ZMQTerminalInteractiveShell.tempfile_image_handler = []

# Use 24bit colors instead of 256 colors in prompt highlighting. If your
#  terminal supports true color, the following command should print 'TRUECOLOR'
#  in orange: printf "\x1b[38;2;255;100;0mTRUECOLOR\x1b[0m\n"
#c.ZMQTerminalInteractiveShell.true_color = False

# Whether to use the kernel's is_complete message handling. If False, then the
#  frontend will use its own is_complete handler.
#c.ZMQTerminalInteractiveShell.use_kernel_is_complete = True

# ------------------------------------------------------------------------------
# KernelManager(ConnectionFileMixin) configuration
# ------------------------------------------------------------------------------

# Manages a single kernel in a subprocess on this host.
#
#  This version starts kernels with Popen.

# Should we autorestart the kernel if it dies.
#c.KernelManager.autorestart = True

# DEPRECATED: Use kernel_name instead.
#
#  The Popen Command to launch the kernel. Override this if you have a custom
#  kernel. If kernel_cmd is specified in a configuration file, Jupyter does not
#  pass any arguments to the kernel, because it cannot make any assumptions about
#  the arguments that the kernel understands. In particular, this means that the
#  kernel does not receive the option --debug if it given on the Jupyter command
#  line.
#c.KernelManager.kernel_cmd = []

# Time to wait for a kernel to terminate before killing it, in seconds.
#c.KernelManager.shutdown_wait_time = 5.0

# ------------------------------------------------------------------------------
# KernelRestarter(LoggingConfigurable) configuration
# ------------------------------------------------------------------------------

# Monitor and autorestart a kernel.

# Whether to include every poll event in debugging output.
#
#  Has to be set explicitly, because there will be *a lot* of output.
#c.KernelRestarter.debug = False

# Whether to choose new random ports when restarting before the kernel is alive.
#c.KernelRestarter.random_ports_until_alive = True

# The number of consecutive autorestarts before the kernel is presumed dead.
#c.KernelRestarter.restart_limit = 5

# Kernel heartbeat interval in seconds.
#c.KernelRestarter.time_to_dead = 3.0

# ------------------------------------------------------------------------------
# Session(Configurable) configuration
# ------------------------------------------------------------------------------

# Object for handling serialization and sending of messages.
#
#  The Session object handles building messages and sending them with ZMQ sockets
#  or ZMQStream objects.  Objects can communicate with each other over the
#  network via Session objects, and only need to work with the dict-based IPython
#  message spec. The Session will handle serialization/deserialization, security,
#  and metadata.
#
#  Sessions support configurable serialization via packer/unpacker traits, and
#  signing with HMAC digests via the key/keyfile traits.
#
#  Parameters ----------
#
#  debug : bool
#      whether to trigger extra debugging statements
#  packer/unpacker : str : 'json', 'pickle' or import_string
#      importstrings for methods to serialize message parts.  If just
#      'json' or 'pickle', predefined JSON and pickle packers will be used.
#      Otherwise, the entire importstring must be used.
#
#      The functions must accept at least valid JSON input, and output *bytes*.
#
#      For example, to use msgpack:
#      packer = 'msgpack.packb', unpacker='msgpack.unpackb'
#  pack/unpack : callables
#      You can also set the pack/unpack callables for serialization directly.
#  session : bytes
#      the ID of this Session object.  The default is to generate a new UUID.
#  username : unicode
#      username added to message headers.  The default is to ask the OS.
#  key : bytes
#      The key used to initialize an HMAC signature.  If unset, messages
#      will not be signed or checked.
#  keyfile : filepath
#      The file containing a key.  If this is set, `key` will be initialized
#      to the contents of the file.

# Threshold (in bytes) beyond which an object's buffer should be extracted to
#  avoid pickling.
#c.Session.buffer_threshold = 1024

# Whether to check PID to protect against calls after fork.
#
#  This check can be disabled if fork-safety is handled elsewhere.
#c.Session.check_pid = True

# Threshold (in bytes) beyond which a buffer should be sent without copying.
#c.Session.copy_threshold = 65536

# Debug output in the Session
#c.Session.debug = False

# The maximum number of digests to remember.
#
#  The digest history will be culled when it exceeds this value.
#c.Session.digest_history_size = 65536

# The maximum number of items for a container to be introspected for custom
#  serialization. Containers larger than this are pickled outright.
#c.Session.item_threshold = 64

# execution key, for signing messages.
#c.Session.key = b''

# path to file containing execution key.
#c.Session.keyfile = ''

# Metadata dictionary, which serves as the default top-level metadata dict for
#  each message.
#c.Session.metadata = {}

# The name of the packer for serializing messages. Should be one of 'json',
#  'pickle', or an import name for a custom callable serializer.
#c.Session.packer = 'json'

# The UUID identifying this session.
#c.Session.session = ''

# The digest scheme used to construct the message signatures. Must have the form
#  'hmac-HASH'.
#c.Session.signature_scheme = 'hmac-sha256'

# The name of the unpacker for unserializing messages. Only used with custom
#  functions for `packer`.
#c.Session.unpacker = 'json'

# Username for the Session. Default is your system username.
#c.Session.username = 'carlosgilgonzalez'
